### ISSUE TITLE
Error thrown `nuxt instance unavailable`, when use `useState` after `useFetch` in composable

### ISSUE THREAD

**Issue by ItsuoSano** [nuxt/framework#5238]
### Environment

------------------------------
- Operating System: `Darwin`
- Node Version:     `v16.13.0`
- Nuxt Version:     `3.0.0-rc.3-27538180.cad4edd`
- Package Manager:  `npm@8.7.0`
- Builder:          `vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------

### Reproduction

https://stackblitz.com/edit/github-cibhp2-szhe5i?file=composables%2Ffetch-data.ts

``` typescript
export const useStateBeforeUseFetch = async () => {
  const { data } = await useFetch('/api/test');
  const state = useState('test');
};
```

### Describe the bug

when I use `useState` after `async useFetch` , error was thrown `nuxt instance unavailable`

### Logs

```shell
[nitro] [dev] [unhandledRejection] Error: nuxt instance unavailable
    at Module.useNuxtApp (file:///home/projects/github-cibhp2-szhe5i/.nuxt/dist/server/server.mjs:408:13)
    at Module.useState (file:///home/projects/github-cibhp2-szhe5i/.nuxt/dist/server/server.mjs:948:38)
    at Module.useStateBeforeUseFetch (file:///home/projects/github-cibhp2-szhe5i/.nuxt/dist/server/server.mjs:3089:39)
```
```

**Comment by igorbelikov** [comment]
+

**Comment by danielroe** [comment]
composables should be run syncronously in most places. (We do some special magic within the body of middleware, and `<script setup>` that makes it possible to mix them in the body of the function, but we don't transform your own composables.)

**Comment by ilyaDegtyarenko** [comment]
> composables should be run syncronously in most places. (We do some special magic within the body of middleware, and `<script setup>` that makes it possible to mix them in the body of the function, but we don't transform your own composables.)

I'm getting this error when refreshing a page after waiting for some idle time on a page where using useAppFetch I'm waiting for data in a <script setup

useAppFetch is my compossable where I wrapped useFetch and added options there, also catch errors and validate token

**Comment by danielroe** [comment]
@ilyaDegtyarenko Do you have some example code?

**Comment by ilyaDegtyarenko** [comment]
> @ilyaDegtyarenko Do you have some example code?

Yes, sure

_**pages/page.vue**_
```
<script
    setup
    lang="ts"
>
    import {useUserStore} from '~/stores/user'

    const {data, pending, refresh} = await useNuxtApp().$api.operator.paginate({
            projectId: userStore.currentProject.id,
            page: page.value
        })
...
```

_**plugins/apiServicePlugin.ts -> operator()**_
```
import {useAppFetch} from '#imports'

export default {
    paginate({projectId, page}: OperatorPaginate) {
        return useAppFetch(`/project/${projectId}/operator`, {
            params: {page},
        })
    },
...
}
```

_**composables/useAppFetch.ts**_
```
import {useAppStore} from '~/stores/app'
import {useUserStore} from '~/stores/user'
import {useLang} from '~/composables/useLang'
import {validate} from '~/utils/jwt'
import {setAuthHeader} from '~/utils/setAuthHeader'
import {UseFetchOptions} from '#app'
import {_AsyncData} from '#app/composables/asyncData'
import {FetchContext, FetchRequest, FetchResponse} from 'ohmyfetch'

export const useAppFetch = async (request: FetchRequest, options: UseFetchOptions<unknown> = {}) => {
    options.baseURL = useRuntimeConfig().apiUrl

    !options.headers && (options.headers = {})

    options.headers['Accept-Language'] = useLang().language.value

    const token = useCookie<string>('token')

    if (validate(token.value)) {
        setAuthHeader(options, token.value)
    } else if (!('Without-Token-Validation' in options.headers)) {
        const refreshToken = useCookie<string>('refreshToken')

        if (!refreshToken.value) {
            useUserStore().logout()

            return
        }

        const {data, error} = await useNuxtApp().$api.auth.refreshToken({refreshToken: refreshToken.value})

        if (error.value) {
            if ([403, 404].includes(error.value.response.status)) {
                useUserStore().logout()

                return
            }

            useAppStore().setError({
                message: `[useAppFetch]: При спробі оновлення токена ` + error.value.response.statusText,
                code: error.value.response.status,
                comment: error.value.data,
            })

            return
        } else {
            token.value = data.value['access_token']
            refreshToken.value = data.value['refresh_token']

            setAuthHeader(options, token.value)
        }
    }

    options.onResponseError = async ({response}: FetchContext & { response: FetchResponse<ResponseType> }): Promise<void> => {
        if ([401, 419].includes(response.status)) {
            useUserStore().logout()
        }
    }

    return useFetch(request, options) as Promise<_AsyncData<any, any>>
}
```

**Comment by danielroe** [comment]
You should get the nuxt instance before any async operations. And I would suggest that you not use plugins in this context, but expose a shared utility so you don't need to call `useNuxtApp` at all to access it.

**Comment by ilyaDegtyarenko** [comment]
> You should get the nuxt instance before any async operations. And I would suggest that you not use plugins in this context, but expose a shared utility so you don't need to call `useNuxtApp` at all to access it.

I turned the plugin into an utility file that I import into the page component, but I still get an error when refreshing the page after a while.

**~/utils/api.ts**
```
import OperatorService from '~/services/operator.service'

export default {
    operator: OperatorService
}
```

**page component**
```
import api from '~/utils/api'

const {data, pending, refresh} = await api.operator.paginate()
```

**Error**
```
[Vue warn]: Unhandled error during execution of setup function 
  at <Operator onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< undefined > key="/p/4/setting/operator" >
[nuxt] [request error] nuxt instance unavailable
```

When I comment on this line, the page opens
`// const {data, pending, refresh} = await api.operator.paginate()
`

**Comment by ilyaDegtyarenko** [comment]
And

Where can I make a handshake request before enabling middlewares? With the placement of the token in cookie for further inquiries.


In the server middleware or plugin, which is better?
If I trying to do this in the app.vue then there it is done after processing the middleware and redirects are performed

**Comment by hermesalvesbr** [comment]
> composables should be run syncronously in most places. (We do some special magic within the body of middleware, and `<script setup>` that makes it possible to mix them in the body of the function, but we don't transform your own composables.)

Sorry, i can´t understand yet. 

1. I can't to use async functions inside composables?
2. I can´t call modules inside composables like [const { getItemById } = useDirectusItems()](https://nuxt-directus.netlify.app/options) ?
3. I can´t get .env variables in composables?

If I can't do any of this, then what should I use in nuxt instead of composables?

**Comment by danielroe** [comment]
Yes, you can call other composables within composables.

Ideally you should not use async functions inside composables. Rather, you would likely want to _return_ an async function so that it can be called within the setup function (or later).

You can get environment variables within composables. But mostly not after an async call.

**Comment by uuf6429** [comment]
I'm having this same problem and on top of @hermesalvesbr, I have to say the composable functionality is extremely frustrating with only a very slim chance of usage.

At this point I don't even know the point behind composables anymore - the documentation says it's to make things easier and avoid boilerplate - but instead we ended up with more complicated and fragile code. On top of this, it seems to me that we are slowly getting composable-only features.

This may sound a bit of a rant (and it is really), I've spent a few days trying to get some functionality to work and instead it's jumping back and forth between component hooks and setup composables.

Anyway, I just want to say that (1) there is real frustration behind these "features" and (2) you can see this from end users. So maybe it's still a good time to go one step back and rethink it a bit.

**Comment by danielroe** [comment]
@uuf6429 Thank you for your thoughts.

Rest assured, making things better and more intuitive is top of my mind.

You may find the following helpful in terms of understanding how Vue composables work, and what their limitations are - particularly note the `Usage Restrictions` section: https://vuejs.org/guide/reusability/composables.html#composables.

**Comment by hermesalvesbr** [comment]
Maybe this can be helpful

https://blog.devgenius.io/ecmascript-2022-is-officially-released-what-should-we-pay-attention-to-5e207ed61a46
